RecoMET/METAlgorithms/src/METSignificance.cc

```diff
@@ -0,0 +1,184 @@
+*/
+//
+// Original Author:  Nathan Mirman (Cornell University)
+//         Created:  Thu May 30 16:39:52 CDT 2013
+//
+//
+
+#include "RecoMET/METAlgorithms/interface/METSignificance.h"
+
+
+metsig::METSignificance::METSignificance(const edm::ParameterSet& iConfig) {
+
+  edm::ParameterSet cfgParams = iConfig.getParameter<edm::ParameterSet>("parameters");
+
+
+  std::string ptResFileName = cfgParams.getParameter<std::string>("ptResFile");
+  std::string phiResFileName = cfgParams.getParameter<std::string>("phiResFile");
+
+  double dRmatch = cfgParams.getParameter<double>("dRMatch");
+  dR2match_ = dRmatch*dRmatch;
+  
+  jetThreshold_ = cfgParams.getParameter<double>("jetThreshold");
+  jetEtas_ = cfgParams.getParameter<std::vector<double> >("jeta");
+  jetParams_ = cfgParams.getParameter<std::vector<double> >("jpar");
+  pjetParams_ = cfgParams.getParameter<std::vector<double> >("pjpar");
+
+
+  edm::FileInPath fpt("CondFormats/JetMETObjects/data/"+ptResFileName);
+  edm::FileInPath fphi("CondFormats/JetMETObjects/data/"+phiResFileName);
+
+  ptRes_  = new JetResolution(fpt.fullPath().c_str(),false);
+  phiRes_ = new JetResolution(fphi.fullPath().c_str(),false);
+
+  fPtEta_ = nullptr;
+  fPhiEta_ = nullptr;
+
+}
+
+metsig::METSignificance::~METSignificance() {
+  delete ptRes_;
+  delete phiRes_;
+  delete fPtEta_;
+  delete fPhiEta_;
+}
+
+
+reco::METCovMatrix
+metsig::METSignificance::getCovariance(const edm::View<reco::Jet>& jets,
+				       const std::vector<reco::Candidate::LorentzVector>& leptons,
+				       const edm::View<reco::Candidate>& pfCandidates) {
+  
+   // metsig covariance
+   double cov_xx = 0;
+   double cov_xy = 0;
+   double cov_yy = 0;
+ 
+   // calculate sumPt
+   double sumPt = 0;
+   for( edm::View<reco::Candidate>::const_iterator cand = pfCandidates.begin();
+         cand != pfCandidates.end(); ++cand){
+      sumPt += cand->pt();
+   }
+
+   //MM & NM :subtraction of lepton will need more study to see which option is optimal
+   //does not signficantly change the results
+
+   // subtract leptons out of sumPt
+   // for ( std::vector<reco::Candidate::LorentzVector>::const_iterator lepton = leptons.begin();
+   //       lepton != leptons.end(); ++lepton ) {
+   //   //sumPt -= lepton->Pt();
+   // }
+   // //protection against unphysical events
+   // if(sumPt<0) sumPt=0;
+
+   // add jets to metsig covariance matrix and subtract them from sumPt
+   for(edm::View<reco::Jet>::const_iterator jet = jets.begin(); jet != jets.end(); ++jet) {
+     
+     // disambiguate jets and leptons
+     if(!cleanJet(*jet, leptons) ) continue;
+
+      double jpt  = jet->pt();
+      double jeta = jet->eta();
+      double feta = std::abs(jeta);
+      double c = std::cos(jet->phi());
+      double s = std::sin(jet->phi());
+
+      // jet energy resolutions
+      double jeta_res = (std::abs(jeta) < 9.9) ? jeta : 9.89; // JetResolutions defined for |eta|<9.9
+      fPtEta_    = ptRes_ -> parameterEta("sigma",jeta_res);
```

Comment:
this should be changed to something like in 
http://cmslxr.fnal.gov/source/RecoMET/METAlgorithms/src/SignAlgoResolutions.cc?v=CMSSW_7_4_0_pre5#0129
we can't really afford churning through TF1 create-delete cycle, which does completely irrelevant things.
@TaiSakuma please help here, you have done this in the past.